client/src/main/java/org/apache/celeborn/client/write/PushState.java

```diff
@@ -29,42 +27,98 @@
 import org.slf4j.LoggerFactory;
 
 import org.apache.celeborn.common.CelebornConf;
+import org.apache.celeborn.common.network.client.RpcResponseCallback;
 import org.apache.celeborn.common.protocol.PartitionLocation;
+import org.apache.celeborn.common.protocol.message.StatusCode;
 
 public class PushState {
+  class BatchInfo {
+    PartitionLocation partitionLocation;
+    ChannelFuture channelFuture;
+    long pushTime;
+    RpcResponseCallback callback;
+  }
+
   private static final Logger logger = LoggerFactory.getLogger(PushState.class);
 
   private int pushBufferMaxSize;
+  private long pushTimeout;
 
   public final AtomicInteger batchId = new AtomicInteger();
-  public final ConcurrentHashMap<Integer, PartitionLocation> inFlightBatches =
+  private final ConcurrentHashMap<Integer, BatchInfo> inflightBatchInfos =
       new ConcurrentHashMap<>();
-  public final ConcurrentHashMap<Integer, ChannelFuture> futures = new ConcurrentHashMap<>();
   public AtomicReference<IOException> exception = new AtomicReference<>();
 
   public PushState(CelebornConf conf) {
     pushBufferMaxSize = conf.pushBufferMaxSize();
+    pushTimeout = conf.pushDataTimeoutMs();
+  }
+
+  public void addBatch(int batchId, PartitionLocation partitionLocation) {
+    BatchInfo info = inflightBatchInfos.computeIfAbsent(batchId, id -> new BatchInfo());
+    info.partitionLocation = partitionLocation;
+  }
+
+  public synchronized void removeBatch(int batchId) {
+    BatchInfo info = inflightBatchInfos.remove(batchId);
+    if (info.channelFuture != null) {
+      info.channelFuture.cancel(true);
+    }
   }
 
-  public void addFuture(int batchId, ChannelFuture future) {
-    futures.put(batchId, future);
+  public int inflightBatchCount() {
+    return inflightBatchInfos.size();
   }
 
-  public void removeFuture(int batchId) {
-    futures.remove(batchId);
+  public synchronized void failExpiredBatch() {
+    long currentTime = System.currentTimeMillis();
+    inflightBatchInfos
+        .values()
+        .forEach(
+            info -> {
+              if (currentTime - info.pushTime > pushTimeout) {
+                if (info.channelFuture != null) {
+                  info.channelFuture.cancel(true);
+                  info.callback.onFailure(
```

Comment:
I think this callback should not be called here hence the future is cancelled, the StdChannelListener will call its operationComplete and enter the failure logic.